### PR TITLE
 Mejoria componente mask

### PR DESCRIPTION
**Pull request by fdiassilva**
## Useful links
- [Contributing](https://github.com/mercadolibre/fury_andesui-android/blob/develop/CONTRIBUTING.md) has more information and tips for a great pull request. This will give you a nice introduction about how to contribute with Andes.
- [Wiki](https://github.com/mercadolibre/fury_andesui-android/wiki) has all the info you need to start with Andes. Even I heard there's a video there...

## Motivation
Tell us why you are here and why you took time to develop new code for our beloved lib.

## Description
Please describe your intentions with this new code you want to introduce inside Andes UI lib.
Try to answer those questions:
- What is it about?
- How does it work?

## Affected component
Is highly probable that this new code directly affects one (or more?) components inside this lib. Tell us who they are!

## Frontify component link
The UX team behind Andes keeps all the Andes Components inside the Frontify site. Each component has its own section. Let us know the section of the component you want to modify.

## Screenshots
This is a UI library, delight us with some stunning images.

## Checks
Are you sure you followed all those steps? In such case, let's say with me "I solemnly declare that ...":
- [X] I have coded an example inside the Demo App,
- [X] I have added proper tests,
- [ ] I have modified the [Changelog](https://github.com/mercadolibre/fury_andesui-android/blob/develop/CHANGELOG.md) file,
- [ ] I have updated GitHub wiki,
- [ ] I have the explicit approval of one or more members of the UX Team,
- [ ] I have updated the version in gradle.properties file.

## Change type
This is easy, let us know what this code is about:
- [ ] This code adds a new component
- [X] This code includes improvements to an existent component
- [ ] This code improves or modifies ONLY the Demo App.

## Check common errors
Whether you are an Android or iOS developer and if you're still there then we'll give you a present:
https://proandroiddev.com/how-to-maximize-androids-ui-reusability-5-common-mistakes-cb2571216a9f

### CODE REVIEW

**Comment by furyGorgoryAdmin** [comment]
### Informative Menu <h2>  ### 

### Use any of these tags in a pull request description or leave a comment ### 
| Tag | Description |
|---|---|
| **[canary]**| Creates a build that can only be deployed with canary strategy. | 
| **[build]** |  Creates version or build from a branch that has been tagged as "no releasable". | 
| **[name:version-name]** |Specify the name of the build. For example: [build] [name:1.2.3-name] or [canary] [name:1.2.3-name].|
| **[major]** | To automatically increase the first position of the version. For example: [build] [major] or [canary] [major].|
| **[minor]** | To automatically increase the second position of the version. For example: [build] [minor] or [canary] [minor]. |
| **[patch]** | To automatically increase the third position of the version. For example: [build] [patch] or [canary] [patch].|  

[For more information about ci-cd](http://docs.furycloud.io/?lang=esp&section=businessassurance)
[For more information about semver](https://semver.org/#summary)

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt [hunk 1]
```diff
@@ -363,7 +367,7 @@ class AndesTextfield : ConstraintLayout {
             counterComponent.visibility = View.VISIBLE
             counterComponent.setTextSize(TypedValue.COMPLEX_UNIT_PX, config.counterSize)
             counterComponent.text = resources.getString(R.string.andes_textfield_counter_text, 0, config.counterLength)
-            textComponent.filters = arrayOf<InputFilter>(InputFilter.LengthFilter(config.counterLength))
+            configTextComponentLength(config)
```
Comment (samuelsainz) [inline components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt:370]: Esto se ejecuta solo si `config.counterLength != 0`. La mask debería definir también el largo del texto. Tiene sentido que se puedan setear ambas cosas?

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt [hunk 2]
```diff
@@ -385,6 +389,17 @@ class AndesTextfield : ConstraintLayout {
         })
     }
 
+    /**
+     * config text length text component
+     */
+    private fun configTextComponentLength(config: AndesTextfieldConfiguration) {
```
Comment (samuelsainz) [inline components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt:395]:
```suggestion
    private fun configTextComponentLengthFilter(config: AndesTextfieldConfiguration) {
```

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt (hunk 1 above)
Comment (fdiassilva) [reply components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt]: Si no desea que aparezca el contador, podemos configurar 0, tenemos un caso en el que el contador solo aparece después de escribir algo y lo controlamos ingresando counter = 0

**Comment by nesmarsollie** [comment]
Estuve revisando toda la lógica de masks porque estoy necesitando algo muy personalizado sobre este componente.
Sin ánimos de criticar, pero me parece que el uso de mask acopla el componente a cosas muy puntuales y cuando esas cosas puntuales fallan , el problema que tenemos es que es muy difícil salir con una solución rápida.

Mi punto de vista sobre mask es que no debería existir así tal cual como lo estamos planteando, sino mas bien, que AndesEdittext podría exponer el watcher y el filter del EditText, y se podrían generar extensiones de kotlin que permitan configurar mask desde afuera del componente, ambos valores con la funcionalidad necesaria. 

De esa forma desacoplamos el componente a mask, y de paso podemos usar mask cuando aplique o sino aplicar otras validaciones con extensiones nuevas parecido a mask pero con controles y funcionalidades diferentes.

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/maskTextField/TextFieldMaskWatcher.kt [hunk 3]
```diff
@@ -34,6 +35,35 @@ class TextFieldMaskWatcher(private var mask: String = "", private var textChange
...
 
+    /**
+     * Format text with mask
+     * @param text without mask
+     */
+    fun getTextWithMask(text: String?): String {
+
+        val editable = SpannableStringBuilder("")
+
+        if (text != null) {
```
Comment (samuelsainz) [inline components/src/main/java/com/mercadolibre/android/andesui/textfield/maskTextField/TextFieldMaskWatcher.kt:46]: Acá estaría bueno usar el let 

text?.let {
     if (mask.isEmpty()) {
          return it.toString()
     }
   
     for (char in it) {
     ....

Comment (fdiassilva) [reply components/src/main/java/com/mercadolibre/android/andesui/textfield/maskTextField/TextFieldMaskWatcher.kt]: Done!

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt [hunk 4]
```diff
@@ -41,7 +40,11 @@ class AndesTextfield : ConstraintLayout {
     var text: String?
         get() = textComponent.text.toString()
         set(value) {
-            textComponent.setText(value)
+            if (maskWatcher != null) {
+                textComponent.setText(maskWatcher!!.getTextWithMask(value))
+            } else {
+                textComponent.setText(value)
+            }
```
Comment (nesmarsollie) [inline components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt:47]: Esto puede generar NPE. Existen 2 alternativas:

```
val mw = maskWatcher
if (mw != null) {
    textComponent.setText(mw.getTextWithMask(value))
} else {
    textComponent.setText(value)
}
```
O esta otra que a mi me gusta mas :

`
textComponent.setText(maskWatcher?.getTextWithMask(value)?:value)
`

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt [hunk 5]
```diff
@@ -377,22 +381,46 @@ class AndesTextfield : ConstraintLayout {
...
      */
-    private fun cleanText(text: String): String? {
+    private fun removeMaskCharsText(text: String?): String {
+        val textToClean = text ?: ""
         return if (maskWatcher == null) {
-            text
+            textToClean
         } else {
-            maskWatcher?.cleanMask(text)
+            maskWatcher?.cleanMask(textToClean)!!
```
Comment (nesmarsollie) [inline components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt:423]: hay que evitar NPE, el !! se podria evitar usando un variable local

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt [hunk 6]
```diff
@@ -557,22 +585,30 @@ class AndesTextfield : ConstraintLayout {
...
+    ) {
 
         if (maskWatcher == null) {
-            maskWatcher = TextFieldMaskWatcher(mask.mask, listener)
+            maskWatcher = TextFieldMaskWatcher(mask, listener)
             textComponent.addTextChangedListener(maskWatcher)
         }
 
-        maskWatcher!!.setMask(mask.mask)
+        maskWatcher!!.setMask(mask)
```
Comment (nesmarsollie) [inline components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt:603]: Hay que reescribir este codigo para evitar el uso de !! y la posibilidad de generar un NPE

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt [hunk 7]
```diff
@@ -557,22 +585,30 @@ class AndesTextfield : ConstraintLayout {
...
 
-        counter = maskWatcher?.getMaxLength()!!
-        textComponent.filters = arrayOf<InputFilter>(InputFilter.LengthFilter((mask.mask.length)))
+        if (maskWatcher?.getLengthWithoutMask() != null && maskWatcher?.getLengthWithoutMask() != 0) {
```
Comment (nesmarsollie) [inline components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt:605]: val length = maskWatcher?.getLengthWithoutMask() ?: 0
if(length > 0) {

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt [hunk 8]
```diff
@@ -557,22 +585,30 @@ class AndesTextfield : ConstraintLayout {
...
+            counter = maskWatcher?.getLengthWithoutMask()!!
```
Comment (nesmarsollie) [inline components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt:606]: counter = length

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt [hunk 9]
```diff
@@ -557,22 +585,30 @@ class AndesTextfield : ConstraintLayout {
...
+            configTextComponentLengthFilter(createConfig())
+        }
 
-        if (mask.digits != null && mask.digits.isNotEmpty()) {
-            textComponent.keyListener = DigitsKeyListener.getInstance(mask.digits)
+        if (digits != null && digits.isNotEmpty()) {
```
Comment (nesmarsollie) [inline components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt:610]: digits.takeIf{ ! it.isNullOrEmpty() }?.let {
     textComponent.keyListener = DigitsKeyListener.getInstance(it)
}

File: demoapp/src/main/java/com/mercadolibre/android/andesui/demoapp/feature/TextfieldShowcaseActivity.kt [hunk 10]
```diff
@@ -146,8 +145,7 @@ class TextfieldShowcaseActivity : AppCompatActivity() {
                 }
 
                 if (mask.text.isNotEmpty()) {
-                    val fieldMask = TextFieldMask(mask = mask.text.toString())
-                    textfield.setTextFieldMask(fieldMask, null)
+                    textfield.setTextFieldMask(mask.text.toString())
                 }
```
Comment (nesmarsollie) [inline demoapp/src/main/java/com/mercadolibre/android/andesui/demoapp/feature/TextfieldShowcaseActivity.kt:149]: Mask es una variable de la cual no tenemos control, en el medio del if podria cambiar su valor.

mask.text.takeIf{ it.isNotEmpty } {
                        textfield.setTextFieldMask(it.toString())
}

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt (hunk 4 above)
Comment (fdiassilva) [reply components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt:47]: textComponent.setText(maskWatcher?.getTextWithMask(value)?:value)

es muito mejor!! gracias!

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt (hunk 6 above)
Comment (fdiassilva) [reply components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt:603]: Nesse caso verifico um pouco acima se nos es Null

Comment (nesmarsollie) [reply components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt:599]: Es correcto, e intuitivamente te diria que no hay problemas.
En android, mas que nada en kotlin, debemos intentar prevenir que race conditions generen errores, dado que muchos procesos pueden actuar en forma simultanea, llamando la misma funcion, podria darse que entre que se verifica y que se use, el valor haya cambiado, y genere NPE. 
Kotlin hizo un paso mas alla, dandonos las herramientas necesarias para evitarlo.

El !! es algo que hay que evitar siempre, diria yo que es un must. Y la forma de evitarlo es realmente sencilla, simplemente creando variables locales evitamos que por algun race condition se pueda generar algun problema.

A simple vista, el codigo hoy por hoy funcionaria perfecto, pero en el futuro las cosas pueden cambiar y podriamos tener problemas.

Mi consejo, y no solo para este revire, es que elimines cualquier uso de !! haciendo el codigo mas seguro.

```
     val currentMask = maskWatcher ?: TextFieldMaskWatcher(mask, listener).also {
              textComponent.addTextChangedListener(it)
              maskWatcher = it
    }

    // A partir de aca usarias para este metodo currentMask porque es seguro usarlo, y es no nulo
    currentMask.setMask(mask)

```

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt [hunk 11]
```diff
@@ -377,22 +377,46 @@ class AndesTextfield : ConstraintLayout {
...
+    /**
+     * return actual text length
+     */
+    private fun getTextLength(): Int {
+
+        return if (text == null || text!!.isEmpty()) {
+            0
+        } else {
+            removeMaskCharsText(text).length
+        }
```
Comment (jiurchuk) [inline components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt:408]: Podemos invertir este if? Creo que es un poco dificil de leer

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt [hunk 12]
```diff
@@ -377,22 +377,46 @@ class AndesTextfield : ConstraintLayout {
...
             }
         })
     }
 
+    /**
+     * config text length text component
+     */
+    private fun configTextComponentLengthFilter(config: AndesTextfieldConfiguration) {
+        if (maskWatcher != null && maskWatcher!!.getMaxLength() != 0) {
+            textComponent.filters = arrayOf<InputFilter>(InputFilter.LengthFilter(maskWatcher!!.getMaxLength()))
```
Comment (jiurchuk) [inline components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt:393]: Qué vuelta le podemos encontrar a no tener que usar el double bang (`!!`) cada vez que usamos el `maskWatcher`?

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt [hunk 13]
```diff
@@ -377,22 +377,46 @@ class AndesTextfield : ConstraintLayout {
...
      */
-    private fun cleanText(text: String): String? {
+    private fun removeMaskCharsText(text: String?): String {
+        val textToClean = text ?: ""
         return if (maskWatcher == null) {
-            text
+            textToClean
         } else {
-            maskWatcher?.cleanMask(text)
+            return maskWatcher?.cleanMask(textToClean) ?: ""
```
Comment (jiurchuk) [inline components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt:419]: Podemos pasar el `""` a una constante estilo `EMPTY_STRING` ?

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt [hunk 14]
```diff
@@ -557,22 +581,31 @@ class AndesTextfield : ConstraintLayout {
...
      * Set mask in text field.
-     * @param mask mask pattern ##-###-######-# digits sample 0123456789-
-     */
-    fun setTextFieldMask(mask: TextFieldMask, listener: TextFieldMaskWatcher.OnTextChange?) {
+     * @param mask pattern ##-###-######-# digits sample 0123456789-
+     * @param digits sample 0123456789-
+     * @param listener is listener changeText
+     */
+    fun setTextFieldMask(
+        mask: String = "",
```
Comment (jiurchuk) [inline components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt:589]: Constante

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/maskTextField/TextFieldMaskWatcher.kt [hunk 15]
```diff
@@ -34,6 +35,34 @@ class TextFieldMaskWatcher(private var mask: String = "", private var textChange
...
+                    if (mask[editableLength] != GENERIC_CHAR) {
+                        editable.append(mask[editableLength])
+                    } else if (mask[editableLength - 1] != GENERIC_CHAR) {
+                        editable.insert(editableLength - 1, mask, editableLength - 1, editableLength)
+                    }
+                }
+            }
+        }
+        return editable.toString()
+    }
```
Comment (jiurchuk) [inline components/src/main/java/com/mercadolibre/android/andesui/textfield/maskTextField/TextFieldMaskWatcher.kt:64]: En el PR de Néstor no se sugerían mejores formas de hacer esto?

Comment (fdiassilva) [reply components/src/main/java/com/mercadolibre/android/andesui/textfield/maskTextField/TextFieldMaskWatcher.kt:64]: Si vamos implementar y extension na proxima version, correcto ?

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt (hunk 13 above)
Comment (nesmarsollie) [reply components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt]: Ahh, yo personalmente disiento en este sentido. "" es bastante claro y no me obliga a revisar el valor de esa constante EMPTY_STRING.
Las constantes son utiles en ocasiones como:
* Cuando su valor depende del negocio 
* Cuando su valor se reusa varias veces y no es trivial
* Cuando es suficientemente complejo como para escribirlo repetidas veces
* Cuando se usa en varias clases 

En estos casos puntuales, mi punto de vista es que una constante "" solo hace las cosas menos legibles y no aporta nada en concreto.

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt (hunk 12 above)
Comment (nesmarsollie) [reply components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt]: de hecho es un problema, y un error usar !! de esta forma. Solo se debe usar !! cuando realmente lo que buscamos es un crash de aplicacion si es nulo.

La solucion es siempre crear variables locales inmutables con el valor original
puntualmente en este caso se puede usar

val maxLength = maskWatcher?.getMaxLength() ?: 0;

Luego se usa esa constante local para todas las comparaciones siguientes.

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt (hunk 11 above)
Comment (nesmarsollie) [reply components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt:411]: de hecho la funcion removeMaskCharsText permite nulos, por lo que esta de mas el if. 
return removeMaskCharsText(text)

es suficiente, y la funcion removeMasCharsText deberia lidiar correctamente con nulos.

Por otro lado esa validacion no es correcta en el sentido de que no garantiza que la funcion sea llamada con el valor de text vacio o nulo, ya que ela comprobacion inicial del if podria ser no nulo, pero por un race condition podria ser nula cuando entra en el bloque else

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt (hunk 12 above)
Comment (nesmarsollie) [reply components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt]: de hecho es un problema, y un error usar !! de esta forma. Solo se debe usar !! cuando realmente lo que buscamos es un crash de aplicacion si es nulo.

La solucion es siempre crear variables locales inmutables con el valor original
puntualmente en este caso se puede usar

val maxLength = maskWatcher?.getMaxLength() ?: 0;

Luego se usa esa constante local para todas las comparaciones siguientes.

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt [hunk 16]
```diff
@@ -377,22 +377,46 @@ class AndesTextfield : ConstraintLayout {
...
+            textComponent.filters = arrayOf<InputFilter>(InputFilter.LengthFilter(maskWatcher!!.getMaxLength()))
+        } else {
+            textComponent.filters = arrayOf<InputFilter>(InputFilter.LengthFilter(config.counterLength))
+        }
+    }
+
+    /**
+     * return actual text length
+     */
+    private fun getTextLength(): Int {
```
[the remark this answers is not in the document]
Comment (fdiassilva) [reply components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt:402]: Done!

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt (hunk 13 above)
Comment (fdiassilva) [reply components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt]: Movi para constant

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt (hunk 11 above)
Comment (fdiassilva) [reply components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfield.kt:408]: invertir  if ficou mais claro, no refactoring irei seguir las sugerencias de Nestor

**Review by samuelsainz — APPROVED** [review]
@fdiassilva fer, solo faltaría agregar la entrada en el Changelog

**Review by jiurchuk — APPROVED** [review]

**Pull request merged**